clean up WebFoundTextRangeController

colelao opened the pull request:
#### 8b622d23c65b0ac89ff5136dd0c70d416d00150d
<pre>
clean up WebFoundTextRangeController
<a href="https://bugs.webkit.org/show_bug.cgi?id=304348">https://bugs.webkit.org/show_bug.cgi?id=304348</a>
<a href="https://rdar.apple.com/166726139">rdar://166726139</a>

Reviewed by Aditya Keerthi.

Some of the code in findTextRangesForStringMatches was difficult to read.
I went through and modernized and cleaned up a bit.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::createWebFoundTextRange):
(WebKit::canConvertToWebFoundTextRange):
(WebKit::findPdfMatchesInFrame):
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):

Canonical link: <a href="https://commits.webkit.org/304713@main">https://commits.webkit.org/304713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d0e45d5bc1214d5be0b5c53bf98ebbecc79a3df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136329 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144041 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/08136f98-b85d-43ba-ad06-7d9798e71e6c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104241 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/77d35449-5d31-4dc0-9fba-4403f9b9a240) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139274 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6824 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122156 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85074 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0f8be51b-b2e7-4915-96fe-9a10721aac5c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6477 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4135 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4632 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115770 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146785 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8368 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112581 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7031 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112925 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28668 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6402 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118460 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8416 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36517 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8134 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71975 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8356 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8208 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -69,60 +70,82 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
-            return Vector<WebFoundTextRange> { };
-        }).iterator->value;
+#if ENABLE(PDF_PLUGIN)
+static inline Vector<WebFoundTextRange::PDFData> findPdfMatchesInFrame(Frame* frame, const String& string, const OptionSet<FindOptions>& options)
```
PDF is an acronym, `OptionSet<FindOptions>` can be passed by value since it fits in a register.

```suggestion
static inline Vector<WebFoundTextRange::PDFData> findPDFMatchesInFrame(Frame* frame, const String& string, OptionSet<FindOptions> options)
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -69,60 +70,82 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
+    if (findMatches.size() > 0)
+        m_cachedFoundRanges.clear();
+
+    auto validSimpleRanges = findMatches | std::views::filter(canConvertToWebFoundTextRange);
+    auto webFoundTextRanges = validSimpleRanges | std::views::transform(createWebFoundTextRange);
```
Technically, this is an additional iteration over the set of matches.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -69,60 +70,82 @@ WebFoundTextRangeController::WebFoundTextRangeController(WebPage& webPage)
+    };
+
+    HashMap<WebCore::FrameIdentifier, Vector<WebFoundTextRange>> frameMatches;
+    for (const auto& [foundTextRange, simpleRange] : std::views::zip(webFoundTextRanges, validSimpleRanges)) {
```
This is also another iteration over matches. But I guess the real bottleneck is the `FIXME` about creating the `CharacterRange`.

colelao replied:
The `std::views` are lazily evaluated, so this is compiled to be just one iteration over the matches. The actual computation happens only once the `std::view` is used in the for loop.

pxlcoder replied:
Cool!

colelao committed:
clean up WebFoundTextRangeController
https://bugs.webkit.org/show_bug.cgi?id=304348
rdar://166726139

Reviewed by Aditya Keerthi.

Some of the code in findTextRangesForStringMatches was difficult to read.
I went through and modernized and cleaned up a bit.

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::createWebFoundTextRange):
(WebKit::canConvertToWebFoundTextRange):
(WebKit::findPdfMatchesInFrame):
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):

Canonical link: https://commits.webkit.org/304713@main
